internal/execute/tableutil/stream.go

```diff
@@ -0,0 +1,184 @@
+package tableutil
+
+import (
+	"context"
+	"sync/atomic"
+
+	"github.com/influxdata/flux"
+	"github.com/influxdata/flux/codes"
+	"github.com/influxdata/flux/internal/errors"
+)
+
+// SendFunc is a function that will be used to construct a stream.
+type SendFunc func(flux.ColReader)
+
+// Stream will call StreamContext with a background context.
+func Stream(f func(ctx context.Context, fn SendFunc) error) (flux.Table, error) {
+	return StreamContext(context.Background(), f)
+}
+
+// StreamContext will create a table that streams column readers
+// through the flux.Table. This method will return only after
+// the function buffers the first column reader.
+// This first column reader is used to identify the group key
+// and columns for the entire table stream.
+//
+// Implementors using this *must* return at least one table.
+// If the function returns without returning at least one table,
+// then an error will be returned. If the first table that is returned
+// is empty, then this will return an empty table and further buffers
+// will not be used.
+func StreamContext(ctx context.Context, f func(ctx context.Context, fn SendFunc) error) (flux.Table, error) {
+	ctx, cancel := context.WithCancel(ctx)
+	ch := make(chan streamBuffer)
+
+	// Create the send method.
+	send := func(cr flux.ColReader) {
+		select {
+		case ch <- streamBuffer{cr: cr}:
+		case <-ctx.Done():
+			// We could not send the column reader because this was cancelled.
+			cr.Release()
+		}
+	}
+
+	done := make(chan struct{})
+	go func() {
+		defer close(done)
+		defer close(ch)
+		if err := f(ctx, send); err != nil {
+			ch <- streamBuffer{err: err}
+		}
+	}()
+
+	select {
+	case sp := <-ch:
+		cr, err := sp.cr, sp.err
+		if cr == nil {
+			if err == nil {
+				err = errors.New(codes.Internal, "empty table stream")
+			}
+			cancel()
+			return nil, err
+		}
+
+		// Retrieve the group key and columns from the column reader.
+		key, cols := cr.Key(), cr.Cols()
+
+		// If the table is empty, signal to the context
+		// that we are not expecting more tables just
+		// in case the implementor does something wrong.
+		// We also release the column reader since we don't need
+		// it anymore and set it to nil.
+		empty := cr.Len() == 0
+		if empty {
+			cancel()
+			cr.Release()
+			cr = nil
+		}
+		return &streamTable{
+			first:  cr,
+			key:    key,
+			cols:   cols,
+			cancel: cancel,
+			ch:     ch,
+			done:   done,
+			empty:  empty,
+		}, nil
+	case <-ctx.Done():
+		cancel()
+		return nil, ctx.Err()
+	}
+}
+
+// streamBuffer is a column reader or error sent
+// from the streaming function.
+type streamBuffer struct {
+	cr  flux.ColReader
+	err error
+}
+
+// streamTable is an implementation of flux.Table
+// that will stream buffers from a column reader.
+type streamTable struct {
+	used   int32
+	first  flux.ColReader
+	key    flux.GroupKey
+	cols   []flux.ColMeta
+	cancel func()
+	ch     <-chan streamBuffer
+	done   <-chan struct{}
```

Comment:
Who will check this channel? I just did not find it and it is private.

Reply:
It is used to prevent the `Do` method from returning until the goroutine associated with streaming ends.

Reply:
Ah, I confused myself with the syntax. Sorry